File: src/csharp/Foundgine.Core/Execution/ProviderPlan.cs
using Foundgine.Core.Semantic.Planning;

namespace Foundgine.Core.Execution;

/// <summary>
/// Opaque physical plan produced by a provider compiler. The core semantic
/// model never depends on its contents.
/// </summary>
public abstract record ProviderPlan(string Provider)
{
    /// <summary>
    /// Security execution certificate. It is intentionally internal so callers
    /// cannot transplant or forge a certificate by assigning it directly.
    /// </summary>
    public SecurityInvariantProof? SecurityProof { get; internal set; }

    /// <summary>
    /// Provenance inherited from the authorized semantic plan. Provider plans
    /// must not be detached from the authorization decision that produced them.
    /// </summary>
    public SemanticPlanAuthorizationBinding? AuthorizationBinding { get; internal set; }

    internal void BindAuthorization(SemanticPlanAuthorizationBinding binding)
    {
        ArgumentNullException.ThrowIfNull(binding);
        if (AuthorizationBinding is not null && AuthorizationBinding != binding)
            throw new InvalidOperationException(
                "Provider plan authorization provenance cannot be replaced once established.");
        AuthorizationBinding = binding;
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/execution/ProviderPlan.java
package com.foundgine.core.execution;

import com.foundgine.core.semantic.planning.SemanticPlanAuthorizationBinding;
import java.util.Objects;

/** Opaque physical plan produced by a provider compiler. */
public abstract class ProviderPlan {
	private final String provider;
	private SecurityInvariantProof securityProof;
	private SemanticPlanAuthorizationBinding authorizationBinding;

	protected ProviderPlan(String provider) {
		if (provider == null || provider.isBlank()) {
			throw new IllegalArgumentException("Provider is required.");
		}
		this.provider = provider;
	}

	public final String provider() {
		return provider;
	}

	public final SecurityInvariantProof securityProof() {
		return securityProof;
	}

	public final SemanticPlanAuthorizationBinding authorizationBinding() {
		return authorizationBinding;
	}

	void setSecurityProof(SecurityInvariantProof proof) {
		this.securityProof = proof;
	}

	void bindAuthorization(SemanticPlanAuthorizationBinding binding) {
		Objects.requireNonNull(binding, "binding");
		if (authorizationBinding != null && authorizationBinding != binding) {
			throw new IllegalStateException(
					"Provider plan authorization provenance cannot be replaced once established.");
		}
		authorizationBinding = binding;
	}
}
